Adapt to TupE/UnboxedTupE having type [Maybe Exp] -> Exp on GHC HEAD

Currently, `th-desugar` fails to build on the latest GHC HEAD, where the types of `TupE` and `UnboxedTupE` have changed from `[Exp] -> Exp` to `[Maybe Exp] -> Exp` to accommodate tuple sections (per [this commit](https://gitlab.haskell.org/ghc/ghc/commit/cef80c0b9edca3d21b5c762f51dfbab4c5857d8a)):

```
[ 9 of 14] Compiling Language.Haskell.TH.Desugar.Core ( Language/Haskell/TH/Desugar/Core.hs, /home/rgscott/Documents/Hacking/Haskell/staging/dist-newstyle/build/x86_64-linux/ghc-8.9.0.20190703/th-desugar-1.10/build/Language/Haskell/TH/Desugar/Core.o, /home/rgscott/Documents/Hacking/Haskell/staging/dist-newstyle/build/x86_64-linux/ghc-8.9.0.20190703/th-desugar-1.10/build/Language/Haskell/TH/Desugar/Core.dyn_o )

Language/Haskell/TH/Desugar/Core.hs:86:23: error:
    • Couldn't match type ‘Maybe Exp’ with ‘Exp’
      Expected type: [Exp]
        Actual type: [Maybe Exp]
    • In the second argument of ‘mapM’, namely ‘exps’
      In a stmt of a 'do' block: exps' <- mapM dsExp exps
      In the expression:
        do exps' <- mapM dsExp exps
           return $ foldl DAppE (DConE $ tupleDataName (length exps)) exps'
   |
86 |   exps' <- mapM dsExp exps
   |                       ^^^^

Language/Haskell/TH/Desugar/Core.hs:89:75: error:
    • Couldn't match type ‘Maybe Exp’ with ‘Exp’
      Expected type: [Exp]
        Actual type: [Maybe Exp]
    • In the second argument of ‘mapM’, namely ‘exps’
      In the second argument of ‘(<$>)’, namely ‘mapM dsExp exps’
      In the expression:
        foldl DAppE (DConE $ unboxedTupleDataName (length exps))
          <$> mapM dsExp exps
   |
89 |   foldl DAppE (DConE $ unboxedTupleDataName (length exps)) <$> mapM dsExp exps
   |                                                                           ^^^^
```

Given the minimalism of `th-desugar`'s design, it seems like tuple sections ought to be desugared to lambda expressions, as originally suggested in [this issue](https://gitlab.haskell.org/ghc/ghc/issues/15843).

--- a/Language/Haskell/TH/Desugar/Core.hs
+++ b/Language/Haskell/TH/Desugar/Core.hs
@@ -26,6 +26,7 @@ import qualified Control.Monad.Fail as Fail
 import Control.Monad.Zip
 import Control.Monad.Writer hiding (forM_, mapM)
 import Data.Data (Data, Typeable)
+import Data.Either (lefts)
 import Data.Foldable as F hiding (concat, notElem)
 import qualified Data.Map as M
 import Data.Map (Map)
@@ -82,11 +83,8 @@ dsExp (LamCaseE matches) = do
   x <- newUniqueName "x"
   matches' <- dsMatches x matches
   return $ DLamE [x] (DCaseE (DVarE x) matches')
-dsExp (TupE exps) = do
-  exps' <- mapM dsExp exps
-  return $ foldl DAppE (DConE $ tupleDataName (length exps)) exps'
-dsExp (UnboxedTupE exps) =
-  foldl DAppE (DConE $ unboxedTupleDataName (length exps)) <$> mapM dsExp exps
+dsExp (TupE exps) = dsTup tupleDataName exps
+dsExp (UnboxedTupE exps) = dsTup unboxedTupleDataName exps
 dsExp (CondE e1 e2 e3) =
   dsExp (CaseE e1 [ Match (ConP 'True [])  (NormalB e2) []
                   , Match (ConP 'False []) (NormalB e3) [] ])
@@ -244,6 +242,49 @@ dsExp (LabelE str) = return $ DVarE 'fromLabel `DAppTypeE` DLitT (StrTyLit str)
 dsExp (ImplicitParamVarE n) = return $ DVarE 'ip `DAppTypeE` DLitT (StrTyLit n)
 dsExp (MDoE {}) = fail "th-desugar currently does not support RecursiveDo"
 #endif
+
+#if __GLASGOW_HASKELL__ >= 809
+dsTup :: DsMonad q => (Int -> Name) -> [Maybe Exp] -> q DExp
+dsTup = ds_tup
+#else
+dsTup :: DsMonad q => (Int -> Name) -> [Exp]       -> q DExp
+dsTup tuple_data_name = ds_tup tuple_data_name . map Just
+#endif
+
+-- | Desugar a tuple (or tuple section) expression.
+ds_tup :: forall q. DsMonad q
+       => (Int -> Name) -- ^ Compute the 'Name' of a tuple (boxed or unboxed)
+                        --   data constructor from its arity.
+       -> [Maybe Exp]   -- ^ The tuple's subexpressions. 'Nothing' entries
+                        --   denote empty fields in a tuple section.
+       -> q DExp
+ds_tup tuple_data_name mb_exps = do
+  section_exps <- mapM ds_section_exp mb_exps
+  let section_vars = lefts section_exps
+      tup_body     = mk_tup_body section_exps
+  if null section_vars
+     then return tup_body -- If this isn't a tuple section,
+                          -- don't create a lambda.
+     else dsLam (map VarP section_vars) tup_body
+  where
+    -- If dealing with an empty field in a tuple section (Nothing), create a
+    -- unique name and return Left. These names will be used to construct the
+    -- lambda expression that it desugars to.
+    -- (For example, `(,5)` desugars to `\ts -> (,) ts 5`.)
+    --
+    -- If dealing with a tuple subexpression (Just), desugar it and return
+    -- Right.
+    ds_section_exp :: Maybe Exp -> q (Either Name DExp)
+    ds_section_exp = maybe (Left <$> qNewName "ts") (fmap Right . dsExp)
+
+    mk_tup_body :: [Either Name DExp] -> DExp
+    mk_tup_body section_exps =
+      foldl' apply_tup_body (DConE $ tuple_data_name (length section_exps))
+             section_exps
+
+    apply_tup_body :: DExp -> Either Name DExp -> DExp
+    apply_tup_body f (Left n)  = f `DAppE` DVarE n
+    apply_tup_body f (Right e) = f `DAppE` e
 
 -- | Desugar a lambda expression, where the body has already been desugared
 dsLam :: DsMonad q => [Pat] -> DExp -> q DExp

--- a/Test/Run.hs
+++ b/Test/Run.hs
@@ -8,7 +8,7 @@ rae@cs.brynmawr.edu
              RankNTypes, TypeFamilies,
              DataKinds, ConstraintKinds, PolyKinds, MultiParamTypeClasses,
              FlexibleInstances, ExistentialQuantification,
-             ScopedTypeVariables, GADTs, ViewPatterns #-}
+             ScopedTypeVariables, GADTs, ViewPatterns, TupleSections #-}
 {-# OPTIONS -fno-warn-incomplete-patterns -fno-warn-overlapping-patterns
             -fno-warn-unused-matches -fno-warn-type-defaults
             -fno-warn-missing-signatures -fno-warn-unused-do-bind
@@ -135,6 +135,9 @@ tests = test [ "sections" ~: $test1_sections  @=? $(dsSplice test1_sections)
 #if __GLASGOW_HASKELL__ >= 807
              , "implicit_params" ~: $test49_implicit_params @=? $(dsSplice test49_implicit_params)
              , "vka"             ~: $test50_vka             @=? $(dsSplice test50_vka)
+#endif
+#if __GLASGOW_HASKELL__ >= 809
+             , "tuple_sections"  ~: $test51_tuple_sections  @=? $(dsSplice test51_tuple_sections)
 #endif
              ]
 

--- a/Test/Splices.hs
+++ b/Test/Splices.hs
@@ -9,7 +9,8 @@ rae@cs.brynmawr.edu
              ScopedTypeVariables, RankNTypes, TypeFamilies, ImpredicativeTypes,
              DataKinds, PolyKinds, GADTs, MultiParamTypeClasses,
              FunctionalDependencies, FlexibleInstances, StandaloneDeriving,
-             DefaultSignatures, ConstraintKinds, GADTs, ViewPatterns #-}
+             DefaultSignatures, ConstraintKinds, GADTs, ViewPatterns,
+             TupleSections #-}
 
 #if __GLASGOW_HASKELL__ >= 711
 {-# LANGUAGE TypeApplications #-}
@@ -272,6 +273,17 @@ test49_implicit_params = [| let f :: (?x :: Int, ?y :: Int) => (Int, Int)
 test50_vka = [| let hrefl :: (:~~:) @Bool @Bool 'True 'True
                     hrefl = HRefl
                 in hrefl |]
+#endif
+
+#if __GLASGOW_HASKELL__ >= 809
+test51_tuple_sections =
+  [| let f1 :: String -> Char -> (String, Int, Char)
+         f1 = (,5,)
+
+         f2 :: String -> Char -> (# String, Int, Char #)
+         f2 = (#,5,#)
+     in case (#,#) (f1 "a" 'a') (f2 "b" 'b') of
+          (#,#) ((,,) _ a _) ((#,,#) _ b _) -> a + b |]
 #endif
 
 type family TFExpand x
@@ -681,5 +693,8 @@ test_exprs = [ test1_sections
 #if __GLASGOW_HASKELL__ >= 807
              , test49_implicit_params
              , test50_vka
+#endif
+#if __GLASGOW_HASKELL__ >= 809
+             , test51_tuple_sections
 #endif
              ]

